csi/controller_server.go

```diff
@@ -279,20 +279,40 @@ func (cs *ControllerServer) isNodeReady(nodeID string) bool {
 	return false
 }
 
-// publishSharedVolume doesn't need to call volume.Attach since the underlying rwo volume will be controlled by a share-manager
-func (cs *ControllerServer) publishSharedVolume(req *csi.ControllerPublishVolumeRequest, volume *longhornclient.Volume) (*csi.ControllerPublishVolumeResponse, error) {
-	if !volume.Ready || len(volume.Controllers) == 0 || volume.Controllers[0].Endpoint == "" {
-		return nil, status.Errorf(codes.Aborted, "The volume %s is not ready for workloads", req.GetVolumeId())
+// publishVolume sends the actual attach request to the longhorn api and executes the passed waitForResult func
+func (cs *ControllerServer) publishVolume(volume *longhornclient.Volume, nodeID string, waitForResult func() error) (*csi.ControllerPublishVolumeResponse, error) {
+	logrus.Debugf("ControllerPublishVolume: volume %s is ready to be attached, and the requested node is %s", volume.Name, nodeID)
+	input := &longhornclient.AttachInput{
+		HostId:          nodeID,
+		DisableFrontend: false,
 	}
 
-	if !cs.waitForVolumeState(volume.Name, "share available", isVolumeShareAvailable, false, false) {
-		return nil, status.Errorf(codes.DeadlineExceeded, "Failed to wait for volume %v share available", req.GetVolumeId())
+	if _, err := cs.apiClient.Volume.ActionAttach(volume, input); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
 	}
```

Comment:
This part is the most important part of this PR,  right? Previously Longhorn CSI plugin will do nothing but just relies on the share manager controller to automatically attach RWX volumes as long as there are workloads using it, then the field `DisableFrontend` won't be reset.

Reply:
Correct we need to call attach for each volume type, so we can reset the disableFrontend flag via the longhorn api -> manager. There is some new logic in the volume manager to deal with this correctly.